### PR TITLE
Base PageProcessor on WorkProcessor

### PR DESCRIPTION
**Pull request by sopel39**
This is initial (Stage 1) PR for effort to support:
* cross operator lazy pages (starting from source operators)
* cleanup/simplify contract between operators via `WorkProcessor` pipelines
* provide base for further improvements (e.g: on stack rows without `Page` materialization, Graal)

The advantage of cross operator lazy pages is that we can avoid IO when queries are highly selective. This requires that significant processing happens in source stage, but this becomes more and more the case with improvements like CBO ("broadcast joins") or grouped execution.

Keep in mind that this is my side effort so I would like to divide it into smaller chunks. 

Stages are:
**Stage 1**
* base `PageProcessor` on `WorkProcessor`

**Stage 2**
* internally base `ScanFilterAndProject` on `WorkProcessor`. The pipeline would look like follows:
```
split singleton -> [flatMap] -> pages source
                -> [transform] -> page processor 
                -> [transform] -> merge pages
```
or if split is cursor based
```
split singleton -> [flatMap] -> cursor source -> [transform] -> merge pages
```
* internally base `FilterAndProject` on `WorkProcessor`. The pipeline would look like follows:
```
page buffer -> [transform] -> page processor -> [transform] -> [merge pages]
```

**Stage 3**
* create interface for operators that are based on `WorkProcessor` pipelines
* create standarized abstract operator class for operators that internally are based on `WorkProcessor` pipelines
* combine operators that are based on `WorkProcessors` via dedicated "gluing" operator
* base `TopNOperator` on `WorkProcessor` pipelines (fast data exploration!)

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 1]
```diff
@@ -42,16 +42,39 @@ private WorkProcessorUtils() {}
...
                     if (processor.process()) {
                         if (processor.isFinished()) {
                             return endOfData();
                         }
 
-                        return processor.getResult();
+                        return Optional.of(processor.getResult());
                     }
                     else if (processor.isBlocked()) {
                         try {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: here we should throw: blocking work processor cannot be (safely) iterated over (add this as a first commit)

then you don't need the `while(true)` in `yieldingIteratorFrom()`

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: addressed in #10768

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: Thanks. Can you rebase now?

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: done

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 2]
```diff
@@ -84,8 +88,9 @@ public void testFilterNoColumns()
 
         Page inputPage = new Page(createLongSequenceBlock(0, 100));
 
-        PageProcessorOutput output = pageProcessor.process(SESSION, new DriverYieldSignal(), inputPage);
-        assertEquals(output.getRetainedSizeInBytes(), inputPage.getRetainedSizeInBytes());
+        AggregatedMemoryContext memoryContext = newSimpleAggregatedMemoryContext();
```
Comment (mbasmanova) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: These 3 lines get repeated a lot. Perhaps, extract a helper method.

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 3]
```diff
@@ -76,42 +85,95 @@ public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection>
...
+        WorkProcessor<Page> processor = process(session, yieldSignal, memoryContext, WorkProcessor.create(new SingletonPage(memoryContext, page)));
+        return processor.yieldingIterator();
+    }
+
+    public WorkProcessor<Page> process(ConnectorSession session, DriverYieldSignal yieldSignal, AggregatedMemoryContext memoryContext, WorkProcessor<Page> pages)
+    {
+        return pages.flatTransform(new PageFilterAndProject(session, yieldSignal, memoryContext));
+    }
+
+    private class SingletonPage
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: This class looks very generic, maybe you could extract a generic `SingletonWorkProcessor<T>`?

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 4]
```diff
@@ -289,6 +352,17 @@ private ProcessBatchResult processBatch(int batchSize)
         }
     }
 
+    @VisibleForTesting
+    public List<PageProjection> getProjections()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: it is not used in tests so far

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 5]
```diff
@@ -221,21 +282,21 @@ public long getRetainedSizeInBytes()
...
-                return Optional.of(page);
+                return ofResult(page);
             }
         }
 
         private void updateRetainedSize()
         {
             // increment the size only when it is the first reference
-            retainedSizeInBytes = Page.INSTANCE_SIZE + SizeOf.sizeOfObjectArray(page.getChannelCount());
+            AtomicLong retainedSizeInBytes = new AtomicLong(Page.INSTANCE_SIZE + SizeOf.sizeOfObjectArray(page.getChannelCount()));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: Why local variable became an `AtomicLong`?

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 6]
```diff
@@ -76,42 +85,95 @@ public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection>
...
-            if (selectedPositions.size() != page.getPositionCount()) {
-                PositionsPageProcessorIterator pages = new PositionsPageProcessorIterator(session, yieldSignal, page, selectedPositions);
-                return new PageProcessorOutput(pages::getRetainedSizeInBytes, pages);
+            Page page = pageOptional.get();
+
+            // limit the scope of the dictionary ids to just one page
+            dictionarySourceIdFunction.reset();
+
+            if (page.getPositionCount() == 0) {
+                return needsMoreData();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: previously code returned `EMPTY_PAGE_PROCESSOR_OUTPUT` but not code is going to loop until it finds non empty page.

**Comment by sopel39** [comment]
@mbasmanova 

> Does WorkProcessor capture computations that will happen in the future if and when the results of these computations are requested?

WorkProcessor concept gives explicit control to processing elements wether they require more data from sources. Because of that, downstream work processor may choose to not materialize page/blocks until they are needed for computations or stored for later use.
The contract here (for pages produced by source operators) is: `page produced by source might be lazy and this laziness is valid until next page is pooled. If processing element wants to keep such lazy page it needs to assure it is loaded before pooling next page from source`.

> Does this help when upstream operator decides to terminate processing early, in which case "future computation" is never performed?

Yes. If pipeline finishes early, then unloaded lazy pages should never be resolved.

> Does this work across stages or only within a stage? Does this apply only until we hit a point where data needs to be computed in full before next step can start, e.g. build hash table for a join?

This is within a stage, so the most benefit is for queries that do most of the processing in source stage (e.g: broadcast joins, star schema queries). Exploration queries are example of such queries (e.g: find top 100 users with highest salary).

IO laziness across stages is rather difficult and could be ineffective as data could be read multiple times (e.g: after data repartition).

> Why existing pipelining mechanism is not sufficient?

Operators as they are now both accept and produce pages. They are essentially double ended queues. `Driver` glues operators together by moving pages between them (in a black box manner). In such model it is difficult and error-prone to implement laziness for IO bound pages as this requires complex multi-party contract (Driver, operators) on how to handle such pages.

On the other hand `WorkProcessor` provide basic top-down approach where downstream elements are in full control wether to delegate computations to upstream elements. There is no intermittent agent between processing elements. In such model it is easier to provide contract for things like laziness.
Additionally, because processing elements are more tightly coupled it is possible to do further interesting optimizations like avoiding pages at all (e.g: process mutable row object).
`WorkProcessors` are also much more low-level. Therefore it makes components like `MergingPageOutput` more reusable as one can just simply plug instance of such element within a pipeline (`MergingPageOutput` is tricky to use right now as it is in-out dequeue).

**Comment by arhimondr** [comment]
@sopel39 Could you please give a real query example you want to optimize, and explain at what steps do we save IO with WorkProcessor concept

**Comment by sopel39** [comment]
@arhimondr 
example source stage (the one that I wanted to do initially):
```
    ...
      |
TopNOperator
      |
ScanFilterProject
```

But this would work with any pipeline like that:
```
...
     |
InsertAnyFilteringOperatorHere
     |
ScanFilterAndProject
```
where `InsertAnyFilteringOperatorHere` would be anything like: `top N, join, filter, semi join with filter, aggregations with masks, etc`

>  explain at what steps do we save IO with WorkProcessor concept

Basically blocks from filtered pages (regions) won't need to be loaded from disk thus saving IOs

**Comment by arhimondr** [comment]
@sopel39 That makes sense. Thanks for the explanation.

**Comment by arhimondr** [comment]
From what i understand you are trying to change what `Operator` means. Currently `Operator` is an entity that does the actual work (e.g. filters the values and materialized the result. What you are trying to make `Operator` be is some kind of a "transformation rule definition".

If you map this to pseudo code, currently we have something like this

```
Page page = input;
page = filter(page);
page = aggregate(page);
page = topN(page);
```

where `filter`, `aggregate` and `topN` operators each return a materialized page.

From what i understand you are trying to get next semantic

```
transformations = transformations(filter(), aggregate(), topN());
Page page = transformations.apply(input)
```

where `filter` , `aggregate` and `topN` operators don't do the actual work, but just define the way data need to be processed.

What is really confusing here, is that you are trying to implement it using the `bottom-up` approach. Instead of first defining the new `Operator` interface, stubbing it with an existing `materializing` implementation and then reimplementing operators one by one, you are trying to introduce the implementation (that is going to be unused for a while) for particular operators. And than you are going to update the interface.

I know that i made a lot of assumptions in my statement. Feel free to correct me if i'm wrong.  But if my assumption are correct, i would rather go with a `top-down` approach. It seems to be more clear, and introduces less unused code = less confusion during the transition period.

**Comment by sopel39** [comment]
@arhimondr 
```
Page page = input;
page = filter(page);
page = aggregate(page);
page = topN(page);
```
On a very schematic level this is true, but when you consider that operators can yield, block, consume pages, produce pages, finish... then the state machine for operators becomes really complex.

> where filter , aggregate and topN operators don't do the actual work, but just define the way data need to be processed.

they do the actual work. In a way the approach is similar to Java `Stream`, but with a support for yielding and blocking. Each processing element is also in control if it wants to pull from source.

> What is really confusing here, is that you are trying to implement it using the bottom-up approach. Instead of first defining the new Operator interface, stubbing it with an existing materializing implementation and then reimplementing operators one by one, you are trying to introduce the implementation (that is going to be unused for a while) for particular operators. And than you are going to update the interface.

I think bottom up approach makes sense for variety of reasons:
* it's easy to make small steps. No need to extend heavy `Operator` interface yet. This way we gradually extend functionality instead of designing new ways operators interact upfront.
* `WorkProcessor` provides abstraction and allows to organize existing code so that there is an immediate benefit of using it. The best example is `ScanFilterAndProject` operator. It deals with `split, PageSource/CursorProcessor, PageProcessor, MergingPageOutput` and their states all mixed up in operator code. With `WorkProcessor` the complexity of interactions between those components is hidden and each component has to manage it's state only.
* I think this approach is kind of similar to how iterative optimizer develops. It gradually replaces old optimizer, but is also extended with new functionalities on the go (e.g: pattern matching, test framework). `IterativeOptimizer` still extends `PlanOptimizer`.
* When two operators are based on a common model like `WorkProcessor` then operator boundaries are artificial really. Some gluing code needs to be added to merge them, but in principle there is no reason why they couldn't be combined together via mechanism like `WorkProcessor`.

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 7]
```diff
@@ -76,42 +85,95 @@ public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection>
...
-            if (projections.isEmpty()) {
-                return new PageProcessorOutput(() -> calculateRetainedSizeWithoutLoading(page), singletonIterator(Optional.of(new Page(selectedPositions.size()))));
+            Page result = page;
+            page = null;
+            memoryContext.close();
+            return ProcessorState.ofResult(result);
+        }
+    }
+
+    private class PageFilterAndProject
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: If the plan is to use this class in other contexts, I think you will need to be careful about the fact this is not a static class and it updating data that is shared across all executions (auto tuning of the work size).

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java (hunk 5 above)
Comment (dain) [reply presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: I don't see what you are talking about in the code below.

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 8]
```diff
@@ -224,8 +226,9 @@ public void testAdaptiveBatchSize()
         Arrays.fill(slices, Slices.allocate(1024));
         Page inputPage = new Page(createSlicesBlock(slices));
 
-        PageProcessorOutput output = pageProcessor.process(SESSION, new DriverYieldSignal(), inputPage);
-        assertEquals(output.getRetainedSizeInBytes(), inputPage.getRetainedSizeInBytes());
+        AggregatedMemoryContext memoryContext = newSimpleAggregatedMemoryContext();
+        Iterator<Optional<Page>> output = pageProcessor.process(SESSION, new DriverYieldSignal(), memoryContext, inputPage);
+        assertEquals(memoryContext.getBytes(), inputPage.getRetainedSizeInBytes());
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: Can this use `processAndAssertRetainedPageSize`?

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 9]
```diff
@@ -241,8 +244,9 @@ public void testAdaptiveBatchSize()
         Arrays.fill(slices, Slices.allocate(128));
         inputPage = new Page(createSlicesBlock(slices));
 
-        output = pageProcessor.process(SESSION, new DriverYieldSignal(), inputPage);
-        assertEquals(output.getRetainedSizeInBytes(), inputPage.getRetainedSizeInBytes());
+        memoryContext = newSimpleAggregatedMemoryContext();
+        output = pageProcessor.process(SESSION, new DriverYieldSignal(), memoryContext, inputPage);
+        assertEquals(memoryContext.getBytes(), inputPage.getRetainedSizeInBytes());
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: same

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 10]
```diff
@@ -335,7 +339,8 @@ public void testYieldProjection()
         Arrays.fill(slices, Slices.allocate(rows));
         Page inputPage = new Page(createSlicesBlock(slices));
 
-        PageProcessorOutput output = pageProcessor.process(SESSION, yieldSignal, inputPage);
+        AggregatedMemoryContext memoryContext = newSimpleAggregatedMemoryContext();
+        Iterator<Optional<Page>> output = pageProcessor.process(SESSION, yieldSignal, memoryContext, inputPage);
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: maybe add an overload so we can use `processAndAssertRetainedPageSize` for all of these

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 11]
```diff
@@ -258,19 +260,19 @@ public LocalMemoryContext newLocalSystemMemoryContext()
     // caller shouldn't close this context as it's managed by the OperatorContext
     public LocalMemoryContext localUserMemoryContext()
     {
-        return new InternalLocalMemoryContext(operatorMemoryContext.localUserMemoryContext(), memoryFuture);
+        return new UncloseableLocalMemoryContext(new InternalLocalMemoryContext(operatorMemoryContext.localUserMemoryContext(), memoryFuture));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: We have too many delegations now: uncloseable context -> internal context -> local user memory context. Maybe we should add constructor arguments to `InternalLocalMemoryContext` and `InternalAggregatedMemoryContext` that tells whether they should support close or not (and remove the uncloseable contexts).

**Comment by sopel39** [comment]
applied comments, should we go ahead?

**Comment by sopel39** [comment]
Here are results of benchmarks run. First row or each pair is `before` this PR, second row is `after` this row:
```
PageProcessorBenchmark.columnOriented              2               false  varchar  avgt   40   153060,622 ±  3818,776  ns/op
PageProcessorBenchmark.columnOriented              2               false  varchar  avgt   40   147778,219 ±  4045,430  ns/op

PageProcessorBenchmark.columnOriented              2               false   bigint  avgt   40    14664,362 ±   279,818  ns/op
PageProcessorBenchmark.columnOriented              2               false   bigint  avgt   40    14511,990 ±   275,336  ns/op

PageProcessorBenchmark.columnOriented              2                true  varchar  avgt   40     6462,877 ±   154,610  ns/op
PageProcessorBenchmark.columnOriented              2                true  varchar  avgt   40     6915,863 ±   169,583  ns/op

PageProcessorBenchmark.columnOriented              2                true   bigint  avgt   40     6152,696 ±   140,111  ns/op
PageProcessorBenchmark.columnOriented              2                true   bigint  avgt   40     6538,855 ±   139,698  ns/op

PageProcessorBenchmark.columnOriented              4               false  varchar  avgt   40   233291,755 ±  4396,434  ns/op
PageProcessorBenchmark.columnOriented              4               false  varchar  avgt   40   235484,023 ±  6160,487  ns/op

PageProcessorBenchmark.columnOriented              4               false   bigint  avgt   40    23166,022 ±   526,638  ns/op
PageProcessorBenchmark.columnOriented              4               false   bigint  avgt   40    23362,950 ±   885,892  ns/op

PageProcessorBenchmark.columnOriented              4                true  varchar  avgt   40     8946,936 ±   193,765  ns/op
PageProcessorBenchmark.columnOriented              4                true  varchar  avgt   40     8867,049 ±   101,872  ns/op

PageProcessorBenchmark.columnOriented              4                true   bigint  avgt   40     8113,478 ±    82,715  ns/op
PageProcessorBenchmark.columnOriented              4                true   bigint  avgt   40     8445,741 ±    96,674  ns/op

PageProcessorBenchmark.columnOriented              8               false  varchar  avgt   40   347150,648 ±  5035,665  ns/op
PageProcessorBenchmark.columnOriented              8               false  varchar  avgt   40   333053,138 ±  5329,785  ns/op

PageProcessorBenchmark.columnOriented              8               false   bigint  avgt   40    37956,031 ±   406,429  ns/op
PageProcessorBenchmark.columnOriented              8               false   bigint  avgt   40    37089,215 ±   378,897  ns/op

PageProcessorBenchmark.columnOriented              8                true  varchar  avgt   40    15994,497 ±   164,317  ns/op
PageProcessorBenchmark.columnOriented              8                true  varchar  avgt   40    15727,663 ±   175,705  ns/op

PageProcessorBenchmark.columnOriented              8                true   bigint  avgt   40    15074,344 ±   125,965  ns/op
PageProcessorBenchmark.columnOriented              8                true   bigint  avgt   40    14823,883 ±   177,135  ns/op

PageProcessorBenchmark.columnOriented             16               false  varchar  avgt   40   629492,480 ±  8758,860  ns/op
PageProcessorBenchmark.columnOriented             16               false  varchar  avgt   40   611255,465 ±  7714,596  ns/op

PageProcessorBenchmark.columnOriented             16               false   bigint  avgt   40    72275,231 ±   717,128  ns/op
PageProcessorBenchmark.columnOriented             16               false   bigint  avgt   40    69789,653 ±   887,009  ns/op

PageProcessorBenchmark.columnOriented             16                true  varchar  avgt   40    29929,931 ±   259,524  ns/op
PageProcessorBenchmark.columnOriented             16                true  varchar  avgt   40    29423,875 ±   289,078  ns/op

PageProcessorBenchmark.columnOriented             16                true   bigint  avgt   40    28737,468 ±   251,477  ns/op
PageProcessorBenchmark.columnOriented             16                true   bigint  avgt   40    27575,682 ±   381,360  ns/op

PageProcessorBenchmark.columnOriented             32               false  varchar  avgt   40  1196437,666 ± 11283,382  ns/op
PageProcessorBenchmark.columnOriented             32               false  varchar  avgt   40  1180375,343 ± 14964,101  ns/op

PageProcessorBenchmark.columnOriented             32               false   bigint  avgt   40   142970,876 ±  2562,730  ns/op
PageProcessorBenchmark.columnOriented             32               false   bigint  avgt   40   136198,419 ±  1619,942  ns/op

PageProcessorBenchmark.columnOriented             32                true  varchar  avgt   40    60995,726 ±   539,681  ns/op
PageProcessorBenchmark.columnOriented             32                true  varchar  avgt   40    57324,366 ±   829,465  ns/op

PageProcessorBenchmark.columnOriented             32                true   bigint  avgt   40    56825,768 ±   760,177  ns/op
PageProcessorBenchmark.columnOriented             32                true   bigint  avgt   40    53869,292 ±   735,705  ns/op
```

**Comment by sopel39** [comment]
rebased @findepi ?

**Comment by sopel39** [comment]
rebased. What shall we do with it?

@dain could you take another look. It partially streamlines flow in `ScanFilterAndProject` operator and `PageProcessor`. Mainly it removes `PageProcessorOutput` and allows for further streamlining. One example where the approach nicely abstracts logic is "window spill" (https://github.com/starburstdata/presto/commit/acd8cb09483458550f20094dab182805a9478e3e), which would be very difficult to write without `WorkProcessors`

I wanted to add some changes related to input operator positions tracking, but I didn't want to conflict with myself.

**Comment by dain** [comment]
The overall shape and new combined abstractions look good to me.  I did not review it to a final level, because I'm not super familiar with the `WorkProcessor` state machine.  I'm hoping that someone the understands the full scope will do a review.  If not, ping me again, and I'll find some time to dig in.

**Comment by sopel39** [comment]
@dain I think it would be the best if you reviewed the code since you are most familiar with `ScanFilterAndProjectOperator`,`PageProcessor`  and `FilterAndProjectOperator`. The flow is mostly the same as previously, but some things around `PageProcessor` changed. For instance, `LocalMemoryContext` is used for `PageProcessor` memory accounting instead of `LongSupplier retainedSizeInBytesSupplier`. I've also reduced number of retain size calculations improving performance.

I did extensive benchmarks and code up to (raw porting to `WorkProcessor`): https://github.com/prestodb/presto/pull/10767/commits/f806b7729a2f43fc4fd109177a6181605eebadf5 has similar performance as the old code. However it's hard to get consistent results as results across multiple executions of `PageProcessorBenchmark` can differ (therefore I compared best runs).

Later I've reduced number of retain size calculations in https://github.com/prestodb/presto/pull/10767/commits/a5894ccae1f2d888907afb06101cc3c58d66167c which improved `PageProcessor` performance in dictionary cases.

I've also identified problems withing `ScanFilterAndProject` operator itself:
* https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java#L236 and https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java#L264 don't progress computations if yield signal is set
* https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java#L264 can do active waiting until `pageSource` is unblocked.

I could address those issues, but I prefer to do so holistically with more suitable `WorkProcessor` abstraction. Then `ScanFilterAndProject` pipeline (for page providing source) would consist of:
`splits -> pageSource -> input pages -> pages transformed via PageProcessor -> page merger`

**Comment by sopel39** [comment]
rebased, kindly ping @dain  ;)

**Comment by sopel39** [comment]
@nezihyigitbasi Operator memory contexts do synchronous update propagation up to the root context (and then further via `DefaultQueryContext` and `MemoryPool`). I suspect those might be bottlenecks when called too frequently (mostly because of synchronization between operators from multiple queries)

File: presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java [hunk 12]
```diff
@@ -308,6 +301,17 @@ private ProcessBatchResult processBatch(int batchSize)
         }
     }
 
+    @VisibleForTesting
+    public List<PageProjection> getProjections()
+    {
+        return projections;
+    }
+
+    private static boolean isUnloadedLazyBlock(Block block)
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java]: "unloaded" block sounds weird. Maybe we can use `isNotLoadedLazyBlock`, or you can inline this at its call site and add a comment.

File: presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java [hunk 13]
```diff
@@ -183,9 +184,6 @@ public long getRetainedSizeInBytes()
     {
         long retainedSizeInBytes = INSTANCE_SIZE;
         retainedSizeInBytes += pageBuilder.getRetainedSizeInBytes();
-        if (currentInput != null) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java]: Why are we not accounting for `currentInput` anymore?

File: presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java [hunk 14]
```diff
@@ -121,8 +125,7 @@ public void testPartialFilter()
 
         Page inputPage = new Page(createLongSequenceBlock(0, 100));
 
-        PageProcessorOutput output = pageProcessor.process(SESSION, new DriverYieldSignal(), inputPage);
-        assertEquals(output.getRetainedSizeInBytes(), inputPage.getRetainedSizeInBytes());
+        Iterator<Optional<Page>> output = processAndAssertRetainedPageSize(pageProcessor, inputPage);
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/project/TestPageProcessor.java]: I debugged the test `testPartialFilter` and in `processAndAssertRetainedPageSize` method we assert that `assertEquals(memoryContext.getBytes(), 0);`. However, when I look at what's being retained by the `output` instance in that method there is plenty of stuff. So, I am confused. This also ties back to my previous question about what we are tracking with these memory contexts.

**Comment by nezihyigitbasi** [comment]
@sopel39 Thanks for the clarification. At this point this patch LGTM. 

I didn't approve the PR, because @dain told me that he also wants to take a look. So, I will let him approve it. (assigning to him)

**Comment by nezihyigitbasi** [comment]
@dain Can you please take a look at this? This has been out there for quite some time.

**Comment by sopel39** [comment]
Thanks!

**Comment by sopel39** [comment]
Related issue: https://github.com/prestodb/presto/issues/12096

**Pull request merged**